cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -0,0 +1,322 @@
+
+    List<FieldOperationInfo> fieldOperationInfos = new ArrayList<>();
+    for (Operation operation : orderedOperations) {
+        fieldOperationInfos.add(convertToFieldOperationInfo(operation));
+    }
+
+    return fieldOperationInfos;
+  }
+
+  @VisibleForTesting
+  List<Operation> getTopologicallySortedOperations(Set<Operation> operations) {
+    // Map of operation name to the set of Operations which take the output of the given operation as an input.
+    // This map basically represents the adjacency list for operation.
+    Map<String, Set<Operation>> outgoingOperations = new HashMap<>();
+    for (Operation operation : operations) {
+      List<InputField> inputFields = new ArrayList<>();
+      switch (operation.getType()) {
+        case TRANSFORM:
+          TransformOperation transform = (TransformOperation) operation;
+          inputFields.addAll(transform.getInputs());
+          break;
+        case WRITE:
+          WriteOperation write = (WriteOperation) operation;
+          inputFields.addAll(write.getInputs());
+          outgoingOperations.put(operation.getName(), new HashSet<>());
+          break;
+      }
+
+      for (InputField inputField : inputFields) {
+        // Add current operation to the adjacency list of operation represented by origin
+        Set<Operation> outgoings = outgoingOperations.computeIfAbsent(inputField.getOrigin(), k -> new HashSet<>());
+        outgoings.add(operation);
+      }
+    }
+
+    Stack<Operation> stack = new Stack<>();
+    Set<String> visitedOperations = new HashSet<>();
+    for (Operation currentOperation : operations) {
+      String currentOperationName = currentOperation.getName();
+      if (visitedOperations.contains(currentOperationName)) {
+        continue;
+      }
+      topologicalSortUtil(outgoingOperations, currentOperation, visitedOperations, stack);
+    }
+
+    List<Operation> topologicalOrder = new ArrayList<>();
+    while (!stack.empty()) {
+      topologicalOrder.add(stack.pop());
+    }
+    return topologicalOrder;
+  }
+
+  /**
+   * Helper function for performing recursion while finding the topological sorting.
+   */
+  private void topologicalSortUtil(Map<String, Set<Operation>> outgoingOperations, Operation currentOperation,
+                                   Set<String> visitedOperations, Stack<Operation> stack) {
+    String currentOperationName = currentOperation.getName();
+    // Mark the current operation as visited.
+    visitedOperations.add(currentOperationName);
+
+    Set<Operation> outgoings = outgoingOperations.get(currentOperationName);
+    if (outgoings == null) {
+      return;
+    }
+
+    for (Operation outgoing : outgoings) {
+      if (visitedOperations.contains(outgoing.getName())) {
+        continue;
+      }
+      topologicalSortUtil(outgoingOperations, outgoing, visitedOperations, stack);
+    }
+
+    stack.push(currentOperation);
+  }
+
+  private FieldOperationInfo convertToFieldOperationInfo(Operation operation) {
```

Comment:
I think this would fit better as a toFieldOperationInfo() method on the Operation class. You could avoid the switch statement and avoid potential future issues if we add a type but forget to change this method.

Reply:
Since Operation is part of API and FieldOperationInfo is in proto kept in the handler.

Reply:
ok I see.